# Definice a importy

In [156]:
import numpy as np
import matplotlib.pyplot as plt
from tabulate import tabulate

In [ ]:
plt.rcParams['text.latex.preamble']=r"\usepackage{mlmodern}"
plt.rcParams["text.usetex"] = True
plt.rcParams["font.size"] = 11
plt.rcParams["font.family"] = 'mlmodern'
#plt.rcParams["mathtext.fontset"] = "cm"

In [19]:
%set_env QXToken=deleteThisAndPasteYourTokenHere

env: QXToken=deleteThisAndPasteYourTokenHere


In [3]:
from typing import List, Callable
from scipy.optimize import minimize
from scipy.optimize._optimize import OptimizeResult

from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector, Operator, SparsePauliOp
from qiskit.primitives import StatevectorSampler, PrimitiveJob
from qiskit.circuit.library import TwoLocal
from qiskit.transpiler.preset_passmanagers import generate_preset_pass_manager
from qiskit.visualization import plot_histogram
from qiskit_ibm_runtime.fake_provider import FakeSherbrooke
from qiskit_ibm_runtime import Session, EstimatorV2 as Estimator
from qiskit_ibm_runtime import SamplerV2 as Sampler
from qiskit_aer import AerSimulator

Usage: qc.draw('mpl', style = my_style)

In [4]:
my_style = {
"backgroundcolor" : "#111111",
"linecolor" : "#FFFFFF",
"textcolor": "#FFFFFF",
"gatetextcolor": "#FFFFFF" }

In [5]:
import qiskit

qiskit.__version__

'1.1.1'

$$\newcommand{\ket}[1]{\left|{#1}\right\rangle}$$ $$\newcommand{\bra}[1]{\left\langle{#1}\right|}$$
$$\newcommand{\braket}[2]{\left\langle{#1}\middle|{#2}\right\rangle}$$
$$\newcommand{\swch}[3]{\langle{#1} | #2 | {#3} \rangle} $$
$$\newcommand{\d}[]{\mathrm{d}}$$
$$\newcommand{\t}[1]{\mathrm{#1}}$$

# Charmonium Hamiltonian

## Formulace problému

Články:
- *Charmonium: The model*
- *Charmonium: the mass spectrum*
- *Properties of Low-Lying Charmonium States in a Phenomenological Approach*


Efektivní jednočásticový potenciál pro $c$ kvark uvězněný v dvoučásticovém $c \bar c$ nazývaný Cornell potential má podobu:
$$
\hat V = -\frac{\alpha_s}{r} + \sigma r
$$
Nerelativistický Hamiltonián pro $c$ kvark v $c \bar c$ tedy lze psát:
$$
\hat H = 2m_c + \hat T + \hat V = 2m_c -\frac{\nabla^2}{2\mu}+\sigma r -\frac{\alpha_s}{r}
$$
kde $\mu = \tfrac{m_c m_c}{m_c + m_c} = \tfrac{1}{2}m_c$ je redukovaná hmotnost pomocí které lze vyjádřit vnitřní kinetickou energii $c \bar c$, $m_c$ klidová hmotnost $c$ kvarku. Spolu s konstantami $\alpha_s$ a $\sigma$ jsou určeny experimentálně a lze je nalézt v tabulce.

Tabulka 1. Konstanty pro výpočty týkající se $c\bar c$:
$$
\begin{array}{cccccc}
\hline \hline \alpha_s & \sqrt{\sigma} & \mu & m_c & \omega & b \\
\hline 0.4038 & 441.6 MeV & 635 MeV & 1270 MeV &  562.9 MeV & 1.7 \times 10^{-3} MeV^{-1} \\
\hline
\end{array}
$$

Při řešení úlohy se omezíme pouze na stavy $c \bar c$ s $l=0$, tedy $s$ stavy. Hamiltonián v druhém kvantování bude mít při uvážení 3 orbitlů s nejnižší energií tvar:
$$
\hat H_3 = \sum_{N=0}^{2} \swch{m}{\hat H}{n} \hat a^\dagger_{m} \hat a^{\ }_n
$$
Orbitaly $\ket{n}$ volíme jako vlastní $s$ stavy Izotropního harmonického oscilátoru s frekvencí $\omega$. Užíváme soustavu jednotek $\hbar = 1$ a zavádíme charakteristickou délku oscilátoru jako $ b = (\mu \omega)^{-\tfrac{1}{2}} $

Stavy Izotropního Harmonického Oscilátoru (IHO) jsou popsány kvantovými čísly $n,l,m$ jako $\ket{n,l,m}$. Lze snadno přejít do sférické souřadnicové reprezentace a stavy tak vyjádřit pomocí vlnových funkcí $\Psi(r,\theta,\varphi)$:
$$ \Psi(r,\theta,\varphi) = \braket{r,\theta,\varphi\,}{\,n,l,m}$$

Hilbertův prostor tvořený těmito funkcemi lze rozložit na Hilbertův prostor kvadraticky integrabilních funkcí na polopřímce $\mathcal{H}_r$ a Hilbertův prostor kv. integrabilních funkcí na jednotkové kouli $\mathcal{H}_\Omega$. Pak $\mathcal{H} \sim \mathcal{H}_r \otimes \mathcal{H}_\Omega$ a vlnové funkce lze rozložit na součin radiální části a sférické části. 
$$\Psi(r,\theta,\varphi) = R_{n,l}(\xi)~Y_{l m}(\theta,\varphi),$$
kde $Y_{l m}(\theta,\varphi)$ jsou kulové funkce.

Kulové funkce mají obecně tvar:
$$ Y_{l m}(\theta,\varphi)=C_{l m}\;P_{l}^{m}(\cos\theta)\;e^{i m\varphi}, $$
kde $P_l^m$ jsou přidružené Legenderovy polynomy. Pro $l=0$ (a tudíž $m=0$) však platí:
$$ Y_{00}(\theta,\varphi)=\frac{1}{\sqrt{4\pi}} $$
Tudíž pro zde uvažované $s$ stavy (tj. $l=0$) je vlnová funkce $\Psi$ pouze funkcí $r$. Radiální část $R_{n,l}(r)$ vlnové funkce pro IHO má tvar:

$$ R_{n,l}(r)=K_{n l}\;\left(\tfrac{r}{b}\right)^{l}\;e^{-\frac{r^{2}}{2b^2}\;}\;L_{n}^{l+\frac{1}{2}}((\tfrac{r}{b})^{2}) $$
$$  K_{n l}=2 \pi^{-\frac{1}{4}} b^{-\frac{3}{2}}\left(\frac{2^{n+l}n!}{(2n+2l+1)!!}\right)^{\frac{1}{2}} $$
$$L_{n}^{\beta}(z)={\frac{1}{n!}}\;e^{z}z^{-\beta}\;{\frac{d^{n}}{d z^{n}}}\;\left(e^{-z}z^{n+\beta}\right)$$

Pro $ l=0 $ pak:
$$R_n(r) = R_{n,0}(r)=(-1)^{n}{\sqrt{{\frac{2n!}{b^{3}\,\Gamma(n+3/2)}}}}\exp\biggl(-{\frac{r^{2}}{2b^{2}}}\biggr)\,L_{n}^{1/2}\biggl({\frac{r^{2}}{b^{2}}}\biggr)\,, $$

## Maticové elementy

Výpočet  $H_{mn} = \swch{m}{\hat H}{n}$ lze rozdělit na několik částí:

$$H_{mn} = \swch{m}{\hat H}{n} = \swch{m}{\hat T}{n} + \swch{m}{\hat V}{n} = T_{mn} + V_{mn}.$$

Pro kinetickou energii $\hat T$ platí:

$$\hat T = \frac{1}{2\mu} \left( \frac{\d^2}{\d r^2} + \frac{2}{r} \frac{\d}{\d r} \right) = \frac{1}{2\mu} \frac{\d}{\d r} r^2 \frac{\d}{\d r}. $$

Maticový element $T_{mn}$ tak lze vyjádřit pomocí integrálu:

$$\swch{m\,}{\hat T}{\,n}  = \int_{0}^{\infty} \frac{1}{2\mu} \frac{\d}{\d r} r^2 \frac{\d}{\d r} (R_m R_n \, r) \mathrm{d}r. $$ 

Maticový element potenciálu $V_{mn}$ lze rozložit dále. Potenciál $\hat V$ má tvar

$$\hat V = -\frac{\alpha_s}{r} + \sigma r.$$

Maticový element lze tedy dále rozložit na součet maticových elementů operátoru násobení $r$ a $r^{-1}$

$$ \swch{m}{\hat V}{n} = -\alpha_s \swch{m}{r^{-1}}{n} + \sigma \swch{m}{r}{n}.$$

Tyto maticové elementy jsou vyjádřeny integrály přes celý prostor prametrizovaný sférickými souřadnicemi. Jak již bylo naznačeno výše, vlnová funkce IHO pro $l=0$ závisí pouze na $r$. Integrací přes úhlové souřadnice je tedy získán celý prostorový úhel $4\pi$. Tento prostorový úhel je pak vykrácen druhou mocninou kulových funkcí v integrálu. Pro určení maticových elementů $r$ tak stačí pouze integrovat přes radiální souřadnici s přidáním $r^2$ z Jakobiánu sférických souřadnic

$$\swch{m\,}{r}{\,n} = \int_{0}^{2\pi} \int_{0}^{\pi} Y_{00}(\theta,\varphi)^2 \int_{0}^{\infty} R_{m,0}(r) R_{n,0}(r) \, r^{3}\sin\theta\,\mathrm{d}r\,\mathrm{d}\theta\,\mathrm{d}\varphi = \int_0^{4\pi} \tfrac{1}{4\pi} \mathrm{d}\Omega \int_{0}^{\infty} R_m R_n \, r^3 \mathrm{d}r = \int_{0}^{\infty} R_m R_n \, r^3 \mathrm{d}r. $$ 

Obdobně lze postupovat i pro $r^{-1}$

$$\swch{m\,}{r^{-1}}{\,n}  = \int_{0}^{\infty} R_m R_n \, r \mathrm{d}r.$$ 

In [46]:
from sympy.physics.sho import R_nl
from sympy.abc import n, r, nu, l, b, psi
from sympy import Function, Symbol
from sympy import Integral, oo

Knihovna sympy umoňuje symbolické výpočty. Radiální funkce IHO jsou v ní zahrnuty. Pro $l=0$ má tvar:

In [108]:
R_nl(n, 0, b**(-2)/2, r)

sqrt(2**(n + 2)*(b**(-2))**(3/2)*factorial(n)/factorial2(2*n + 1))*exp(-r**2/(2*b**2))*assoc_laguerre(n, 1/2, r**2/b**2)/pi**(1/4)

Integrály s $R_{nl}$ lze počítat následujícím způsobem: 

In [114]:
ksi = 1.7*10**(-3)
Integral(
    R_nl(0,0,ksi**(-2)/2,r)**2*r**3, 
    (r, 0, oo)
).n()*441.6**2

374.077950554970

In [116]:
Integral(
    R_nl(0,0,ksi**(-2)/2,r)**2*r, 
    (r, 0, oo)
).n()*0.408

270.811000102923

### Výpočet $V_{mn} = \swch{m}{\hat V}{n}$

In [161]:
V = np.zeros((3,3))
ksi = 1.66*10**(-3)

for i in range(3):
    for j in range(3):
        i1 = Integral(R_nl(i,0,ksi**(-2)/2,r)*R_nl(j,0,ksi**(-2)/2,r)*r**3, (r, 0, oo)).n()
        i2 = Integral(R_nl(i,0,ksi**(-2)/2,r)*R_nl(j,0,ksi**(-2)/2,r)*r, (r, 0, oo)).n()
        v = (i1*441.6**2 - i2*0.4038)*(-1)**(i+j)
        print("V{}{} = {}".format(i,j,v), end = " ")
        if j==2: print("")
        V[i][j] = v

V00 = 90.7944852958506 V01 = 261.180006749393 V02 = -108.514885275498 
V10 = 261.180006749393 V11 = 319.179482029386 V12 = 316.718147446441 
V20 = -108.514885275498 V21 = 316.718147446441 V22 = 481.318841875256 


In [163]:
print(V)

[[  90.7944853   261.18000675 -108.51488528]
 [ 261.18000675  319.17948203  316.71814745]
 [-108.51488528  316.71814745  481.31884188]]


In [165]:
tab = V
#print(tabulate(tab, tablefmt="latex_booktabs"))

Matice definovaná elementy $\hat V_{mn}$ má tak tvar:
$$
\begin{pmatrix*}[r]
   90.7945 & 261.18  & -108.515 \\
  261.18   & 319.179 &  316.718 \\
 -108.515  & 316.718 &  481.319 \\
\end{pmatrix*}
$$

### Výpočet $T_{mn} = \swch{m}{\hat T}{n}$

### Extra LaTeX



$$\psi_{n l m}(r,\theta,\phi)=N\left(\frac{r}{b}\right)^{l}L_{n}^{l+1/2}(\frac{r}{b})e x p\left(-\frac{r^{2}}{2b^{2}}\right)Y_{l m}(\theta,\phi)$$
$$ |N|^{2}=\frac{2\alpha^{3}n!}{\sqrt{\pi}}\frac{2^{(2(n+l)+1)}}{(2n+2l+1)!}(n+l)!$$



VQE Článek:
$$ \langle r|n\rangle=(-1)^{n}{\sqrt{{\frac{2n!}{b^{3}\,\Gamma(n+3/2)}}}}\exp\biggl(-{\frac{r^{2}}{2b^{2}}}\biggr)\,L_{n}^{1/2}\biggl({\frac{r^{2}}{b^{2}}}\biggr)\,, $$

Tahák:

$$\psi_{n,l,m}(\xi,\theta,\varphi)=R_{n,l}(\xi)~Y_{l m}(\theta,\varphi)$$
$$ R_{n,l}(\xi)=K_{n l}\;\xi^{l}\;e^{-\frac{\xi^{2}}2\;}\;L_{n}^{l+\frac{1}{2}}(\xi^{2}) $$
$$ \xi=r\sqrt{\frac{M\omega}{\hbar}},\ \ \ K_{n l}=\frac{2}{\pi^{\frac{1}{4}}}\left(\frac{M\omega}{\hbar}\right)^{\frac{3}{4}}\left(\frac{2^{n+l}n!}{(2n+2l+1)!!}\right)^{\frac{1}{2}} $$
$$L_{n}^{\beta}(z)={\frac{1}{n!}}\;e^{z}z^{-\beta}\;{\frac{d^{n}}{d z^{n}}}\;\left(e^{-z}z^{n+\beta}\right)$$
$$ {L}_{0}^{\beta}(z)=1,\ {L}_{1}^{\beta}(z)=1+\beta-z $$

Kulové funkce:
$$ Y_{l m}(\theta,\varphi)=C_{l m}\;P_{l}^{m}(\cos\theta)\;e^{i m\varphi}, $$
kde $P_l^m$ jsou přidružené Legenderovy polynomy. Pro $l=0$ (a tudíž $m=0$) však platí:
$$ Y_{00}(\theta,\varphi)=\frac{1}{\sqrt{4\pi}} $$